### DISCUSSION TITLE
xstate with Cloudflare Workers Durable Objects

### DISCUSSION THREAD

**Discussion by nathanclevenger** [statelyai/xstate#3502] [Ideas]
Cloudflare Workers [Durable Objects](https://blog.cloudflare.com/durable-objects-easy-fast-correct-choose-three/) now provide a global unique singleton instance of a class for edge performance with transactional integrity.  Is there any interest in adding official xstate support for Durable Objects with the [transactional storage api](https://developers.cloudflare.com/workers/runtime-apis/durable-objects/#transactional-storage-api)?

This is something we are actively looking to build at @drivly, and we would be happy to contribute here.

**Comment by davidkpiano** [comment]
Hey @nathanclevenger, we've played around with DO months ago (with @gavination) and I'd love to see an example of how you'd use it with XState.

Can you make an example repository with running an XState machine in a DO, and we can start from there? I would be in favor of making an official adapter if we can determine what that should look like.

**Reply by nathanclevenger** [reply]
Will do!  We'll pull together an example or two and post the link here.

**Reply by sam-lippert** [reply]
Hello, our Durable Object xstate engine now has an initial implementation with the machine definition and state stored in transactional storage: https://github.com/drivly/state.do
There are also callback URL extensions added to this machine so that behavior can be defined declaratively.

**Reply by davidkpiano** [reply]
@sam-lippert This looks great! Let me know if there are things we can improve on our side to make the XState + DO integration work more smoothly.
